<!doctype html>
<html>
  <head>
    <meta charset="utf-8">
    <meta http-equiv="X-UA-Compatible" content="chrome=1">
    <title>Ae aegypti-toolset by mshadbolt</title>
      
    <link rel="icon" href="/images/mozzieicon.ico" type="image/ico" />
    <link rel="stylesheet" href="stylesheets/styles.css">
    <script src="https://use.fontawesome.com/0647d4ff24.js"></script>
    <link rel="stylesheet" href="stylesheets/github-light.css">
    <link href="https://fonts.googleapis.com/css?family=Lato" rel="stylesheet">
    <link rel="stylesheet" href="//code.jquery.com/ui/1.12.0/themes/smoothness/jquery-ui.css">
    <meta name="viewport" content="width=device-width">
    <script src="https://ajax.googleapis.com/ajax/libs/jquery/1.12.4/jquery.min.js"></script>
    <!--[if lt IE 9]>
    <script src="//html5shiv.googlecode.com/svn/trunk/html5.js"></script>
    <![endif]-->
  </head>
  <body>
        <div class="wrapper">
          <header>
            <img src="images/mozzie.png" style="width:200px" >
              <h1><span style="font-style: italic">Aedes aegypti</span> Genomics Toolset</h1>
            <p>Set of scripts to aide <span style="font-style: italic; font-weight: bold">Ae. aegypti</span> genomic analyses</p>

            <p class="view"><a href="https://github.com/mshadbolt/Ae_aegypti-toolset">View the Project on GitHub <small>mshadbolt/Ae_aegypti-toolset</small></a></p>


            <ul>
              <li><a href="https://github.com/mshadbolt/Ae_aegypti-toolset/zipball/master">Download <strong>ZIP File</strong></a></li>
              <li><a href="https://github.com/mshadbolt/Ae_aegypti-toolset/tarball/master">Download <strong>TAR Ball</strong></a></li>
              <li><a href="https://github.com/mshadbolt/Ae_aegypti-toolset">View On <strong>GitHub</strong></a></li>
            </ul>
          </header>
            
        <section>
            <h1>
                <a id="aeaegypti-toolset" class="anchor" href="#aeaegypti-toolset" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>
            </h1>

            <h2>R tutorials</h2>
            <ul>
              <li><a href=AccessingVectorbaseinBiomaRt.html>Accessing VectorBase in R with biomaRt</a></li>
              <li><a href=GowindaAssocFile.html>Create a Gene Set file using R for input to Gowinda</a></li>
            </ul>

            <h2>Commandline scripts</h2>
            <div class="accordion">
                <div class="accordion-section">
                    <a class="accordion-section-title" href="#accordion-1">AaegL1/2/3>Juneja VCF Converter</a>

                    <div id="accordion-1" class="accordion-section-content">
                        <p><span style="color: #1a1a1a; font-weight: bold">Description: </span>Converts vcf file aligned to AaegL1/2/3 (superconts) into mapped positions on chromosomes according to <a href="http://dx.doi.org/10.1371/journal.pntd.0002652" target="_blank">Juneja et al. 2014</a>. Supercontigs that were not able to be mapped are left as is. Positions are then sorted by chromosome so chr1-3 come first, followed by unmapped chromosomes.</p>
                        <p><span style="color: #1a1a1a; font-weight: bold">Usage: </span><code>python variantpositions.vcf</code></p>
                        <p><span style="color: #1a1a1a; font-weight: bold">Input:</span> Standard <a href="https://en.wikipedia.org/wiki/Variant_Call_Format">VCF file</a> where header rows commence with ## and the first two columns are CHROM and POS based on positions aligned to AaegL1-3. Also required is the supplemental file outlining mappings between supercontigs and chromosomes called <code>JunejaGeneticAssembly.csv</code></p>
                        <p><span style="color: #1a1a1a; font-weight: bold">Output: </span>Copy of the given VCF file with prefix <code>juneja.</code></p>
                        <p><span style="color: #1a1a1a; font-weight: bold">Download: </span><a href="https://github.com/mshadbolt/Ae_aegypti-toolset/raw/master/AeagyptVCFConverter.py" download>AaegyptVCFConverter.py</a> & <a href="https://github.com/mshadbolt/Ae_aegypti-toolset/raw/master/JunejaGeneticAssembly.csv" download>JunejaGeneticAssembly.csv</a> </p>
                    </div><!--end .accordion-section-content-->
                </div><!--end .accordion-section-->
                <div class="accordion-section">
                    <a class="accordion-section-title" href="#accordion-2">UniProt&BioMart>Gowinda</a>

                    <div id="accordion-2" class="accordion-section-content">
                        <p><span style="color: #1a1a1a; font-weight: bold">Description:</span> Downloads latest <span style="font-style: italic">Ae. aegypti</span> associated <a href="http://geneontology.org/" target="_blank">Gene Ontology</a> terms from UniProt & BioMart using their REST APIs and creates input to the program <a href="https://sourceforge.net/p/gowinda/wiki/Main/">Gowinda</a> (<a href="https://doi.org 10.1093/bioinformatics/bts315">publication</a>)</p>
                        <p><span style="color: #1a1a1a; font-weight: bold">Usage:</span> <code>python UniprotToGowinda.py</code></p>
                        <p><span style="color: #1a1a1a; font-weight: bold">Depends:</span> Requires the python module <code>wget</code> to be installed. Should be as simple as using the command: <code>pip install wget</code> in the terminal if you don't have it.</p>
                        <p><span style="color: #1a1a1a; font-weight: bold">Input:</span> Optional to supply already downloaded files from Uniprot and GO consortium using the flags: </p>
                        <p><code>--gene_file [uniprotoutput.tab.gz]</code></p>
                        <p><code>--obo_file [go.obo] </code></p>
                        <p><span style="color: #1a1a1a; font-weight: bold">Output:</span> Tab-delimited text file entitled <code>UniprotBioMartGOassocGowinda.txt</code> ready for input into Gowinda, eg:               <blockquote>
                                GO:0007005  mitochondrion_organization	AAEL007089 AAEL010831 AAEL001853 AAEL003829 AAEL011606 <br/>
                                GO:0006228	UTP_biosynthetic_process	AAEL003030 AAEL004107 AAEL011098 AAEL012359
                            </blockquote>  
                        </p>
                        <p>Download: <a href="https://github.com/mshadbolt/Ae_aegypti-toolset/raw/master/UniprotBioMartToGowinda.py" download>UniprotToGowinda.py</a></p>
                    </div><!--end .accordion-section-content-->
                </div><!--end .accordion-section-->
            </div><!--end .accordion-->
          
            <br/>
            <br/>
            
          <h2>Web resources</h2>
            <ul>
                <li>
                    <a href="http://www.uniprot.org/uniprot/?query=aedes%20aegypti&fil=organism%3A%22Aedes+aegypti+%28Yellowfever+mosquito%29+%28Culex+aegypti%29+%5B7159%5D%22&sort=score">UniProt entries for <span style="font-style: italic">Ae. aegypti</span></a>
                </li>
                <li>
                    <a href="https://www.vectorbase.org/downloads?field_organism_taxonomy_tid=372&field_download_file_type_tid=All&field_download_file_format_tid=All&field_status_value=Current">VectorBase data downloads for <span style="font-style: italic">Ae. aegypti</span></a>
                </li>
            </ul>
          
        </section>

<!--
<p>On the wishlist:</p>

<ul>
<li>In silico genomic sexing based on Fontaine et al. (2016) <a href="http://dx.doi.org/10.1101/060061">http://dx.doi.org/10.1101/060061</a>
</li>
<li>Convert PLINK output files mapped to supercontigs to mapping to Juneja chromosomes - Then potentially use as input to SNeP -&gt; <a href="https://sourceforge.net/projects/snepnetrends/?source=navbar">https://sourceforge.net/projects/snepnetrends/?source=navbar</a> <a href="http://dx.doi.org/10.3389/fgene.2015.00109">http://dx.doi.org/10.3389/fgene.2015.00109</a>
</li>
<li>R mapping superconts to Juneja chromosomes script</li>
<li>R LD visualisation Shiny app</li>
<li>Updated trimming and stacks scripts (?)</li>
</ul>
-->
     
      
      <footer>
        <p>This project is maintained by <a href="https://github.com/mshadbolt">mshadbolt</a></p>
        <p><a href="https://twitter.com/MazzaShazza" class="twitter-follow-button" data-show-count="false"><i class="fa fa-twitter" aria-hidden="true"></i> Follow @MazzaShazza</a></p>
        <p><small>Hosted on GitHub Pages &mdash; Theme by <a href="https://github.com/orderedlist">orderedlist</a></small></p>
      </footer>
    </div>
    <script src="javascripts/scale.fix.js"></script>
    <script src="javascripts/main.js"></script>
    
  </body>
</html>